src/sys/windows.rs

```diff
@@ -735,6 +736,52 @@ pub(crate) fn from_in6_addr(addr: in6_addr) -> Ipv6Addr {
     Ipv6Addr::from(*unsafe { addr.u.Byte() })
 }
 
+/// This type is not yet in winapi.
+#[repr(C)]
+#[allow(non_camel_case_types)]
+struct sockaddr_un {
+    pub sun_family: sa_family_t,
+    pub sun_path: [u8; 108],
+}
+
+#[allow(unused_unsafe)] // TODO: replace with `unsafe_op_in_unsafe_fn` once stable.
+pub(crate) fn unix_sockaddr(path: &Path) -> io::Result<SockAddr> {
+    unsafe {
+        SockAddr::init(|storage, len| {
+            // Safety: `SockAddr::init` zeros the address, which is a valid
+            // representation.
+            let storage: &mut crate::sys::sockaddr_un = unsafe { &mut *storage.cast() };
+            let len: &mut socklen_t = unsafe { &mut *len };
+
+            let bytes = path
+                .to_str()
+                .ok_or_else(|| {
+                    io::Error::new(io::ErrorKind::InvalidInput, "path must be valid UTF-8")
+                })?
+                .as_bytes();
+
+            if bytes.len() > storage.sun_path.len() {
```

Comment:
This needs the same check as in https://github.com/rust-lang/socket2/blob/1c67209c2eda5708b3093900a59befbc4a981b13/src/sys/unix.rs#L460-L472

If it doesn't support abstract namespaces you can remove that bit. The correct code won't work if `bytes.len() == storage.sun_path.len()` as the string isn't null terminated then.

Reply:
Windows should allow the full sun_path to be filled out without a null terminator.

Reply:
In that case can you add a comment stating that and referring to the Windows docs? I the only docs I read about it said it was Linux compatible (which it still is, but the not null-terminated usage isn't as that is invalid on Unix).

Reply:
I used to manage the team that implemented this in Windows, but it appears that we didn't document this behavior. I'll keep the null terminator requirement for now and we can relax it once I get the documentation fixed.